### PR TITLE
refactor: authz provider implementation and authz users basic implementation  

### PR DESCRIPTION
**Pull request by NicholasBlaskey**
## Description

Refactor the user api both echo and grpc to use the new ```authz``` pattern.

## Test Plan

existing e2e tests + new integration tests

## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.
- [ ] If modifying `/webui/react/src/shared/` verify `make -C webui/react test-shared` passes.

### CODE REVIEW

File: master/internal/api_user.go [hunk 1]
```diff
@@ -207,10 +239,20 @@ func (a *apiServer) PatchUser(
...
+		return nil, err
 	}
+	targetUser := targetFullUser.ToUser()
+	if err = user.AuthZProvider.Get().CanSetUsersDisplayName(*curUser, targetUser); err != nil {
+		if !user.AuthZProvider.Get().CanGetUser(*curUser, targetUser) {
+			return nil, errUserNotFound
+		}
+		return nil, errors.Wrap(grpcutil.ErrPermissionDenied, err.Error())
+	}
+
```
Comment (eecsliu) [inline master/internal/api_user.go:255]: no check for if the user can patch other users?

File: master/internal/config/authz_config.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+func initAuthZTypes() {
+	authZConfigMutex.Lock()
+	defer authZConfigMutex.Unlock()
+
+	if knownAuthZTypes != nil {
+		return
+	}
+
+	knownAuthZTypes = make(map[string]bool)
```
Comment (eecsliu) [inline master/internal/config/authz_config.go:63]: It seems like `knownAuthZTypes` can be remade if this is called more than once? Should be locked down with a once like in the provider

**Review by eecsliu — COMMENTED** [review]
Looks good! Couple of questions below

File: master/internal/config/authz_config.go (hunk 2 above)
Comment (NicholasBlaskey) [reply master/internal/config/authz_config.go:63]: There is a nil check for ```knownAuthZTypes``` above -- which I missed for like a couple minutes wondering how it even worked lol

Comment (eecsliu) [reply master/internal/config/authz_config.go:63]: oh lmao I'm just blind

**Comment by eecsliu** [comment]
Is this dependent on Ilia's authz work? Would you want to merge this into `rbac-user-groups` branch on top of his changes before merging all of the user groups stuff into master together?

**Comment by NicholasBlaskey** [comment]
> Is this dependent on Ilia's authz work? Would you want to merge this into `rbac-user-groups` branch on top of his changes before merging all of the user groups stuff into master together?

Ilia's authz work is already in this PR and I think I would rather to keep commit history cleaner keeping this change separate?  

Also I think I changed my mind about ```CanGetUser``` not returning an error. I am thinking it would be pretty bad UX for a user to like try to do an operation on another user and get not found when they should get a database failure error. I'm going to make ```CanGetUser``` return ```(canViewUser bool, serverError error)``` to try and make it clear that an error shouldn't be returned for denying viewing a user.

**Review by eecsliu — APPROVED** [review]

**Pull request merged**